### ISSUE TITLE
Implementing RV32E on example UPduino-v3.0

### ISSUE THREAD

**Issue by ajngo** [stnolting/neorv32-setups#143]
Hi, I am fairly new to FPGA and RISCV design. I have been trying to get the RV32E extension to work on the example UPduino project. A few months ago I was able to successfully setup the default provided project but unable to get RV32E to run with an example program. Recently I came back to it and redownloaded the example after the current rv32e prebuilt compiler was committed. Now I cannot get the default project to synthesize in Lattice Radiant. It gives an error saying 'neorv32_cpu_cp_cond is not compiled in library neorv32'. I've also tried using an open-source toolchain but I get an error saying 'no such command: ghdl' even though I have the latest version of ghdl installed on ubuntu. Any advice for the issue above and implementing RV32E with an example program would be much appreciated!

**Comment by stnolting** [comment]
Hey there!

> Now I cannot get the default project to synthesize in Lattice Radiant. It gives an error saying 'neorv32_cpu_cp_cond is not compiled in library neorv32'.

Sorry for that! The `Zicond` support is quite new and I think I've forgotten to add that to the example setup. I'll fix/update the Radiant project file.

> I've also tried using an open-source toolchain but I get an error saying 'no such command: ghdl' even though I have the latest version of ghdl installed on ubuntu.

You'll also need yosys. If you like you use this docker image where all required tools are included: https://github.com/stnolting/neorv32/pkgs/container/neorv32%2Fimpl We are using this image for the gitHub auto-build of the open-source setups from this repository.

**Comment by ajngo** [comment]
Thanks for the response! I have gone ahead and downloaded the updated files, but while synthesizing in Lattice Radiant it pauses after the place and route section and says there is a timing error. I've tried to continue with creating the bit file, but when I program the board the LED doesn't flash even though the programmer says it was successful. Do you have any insight into what may be causing the timing error?

**Comment by stnolting** [comment]
Hmm this is the strange. The default Radiant project is working out of the box over here. 🤔

* What version of Radiant are you using?
* Did you change the default setup?
* Can you identify in what part of the processor the timing violation occurs?

**Comment by stnolting** [comment]
> I am using the latest version of Radiant 2023.2

:+1:

Which version of the NEORV32 processor are you using? Also the latest (v1.9.3.3)?

> I'm not sure but I see from the message console that a lot of registers have bit values stuck. I have attached screenshots of the console with the warnings

I think that is fine. Most of these warning are generated by hardware module and ISA extension that were not enabled so their output signals are hardwired to zero.

I have re-synthesized the design and I am getting no timing errors at all. Unfortunately, I do not have a UPduino available right now, so I cannot test it. However, you could try to re-generate the PLL IP. Maybe something is broken there.

Also, please try switching to "Sinplify Pro" as synthesis engine (the LSE is ... tricky ... sometimes):

![grafik](https://github.com/stnolting/neorv32-setups/assets/22944758/e5496958-23f5-4059-bca5-1d1f25ad6110)

**Comment by ajngo** [comment]
Yes I am using the latest version of the NEORV32 processor. I have also tried using the neorv32 submodule in neorv32-setups but that doesn't work either. One thing I have noticed is that after re-generating all IPs the LSE synthesis tool synthesizes with no problem, however the synplify doesn't work no matter what. Also, I have found that when opening the programmer from the Radiant IDE I get a read error with the board. 
<img width="772" alt="read_error" src="https://github.com/stnolting/neorv32-setups/assets/63076423/364999a1-67a7-4fcb-89ff-42e26ebad246">

But if I open the programmer independently it programs successfully, albeit the LED doesn't flash

**Comment by stnolting** [comment]
I do not have a Upduino right now, so I cannot test _that_ setup. However, I am using a ICEbreaker FPGA board featuring the same chip and the setup (using Sinplify) works out of the box.

> however the synplify doesn't work no matter what

Do you get any errors there?

> Also, I have found that when opening the programmer from the Radiant IDE I get a read error with the board.

I think I've see that error before. Did you select the right cable port? Clicking "Detect Cable" in the right top of the programmer should find the correct settings. If not, you could try all four cables/ports manually.

**Comment by ajngo** [comment]
Yes, synplify gives this error
<img width="455" alt="Untitled" src="https://github.com/stnolting/neorv32-setups/assets/63076423/9f04fe45-4b9b-418d-aafb-1cd9bae8cadd">
LSE synthesizes with no errors, but when I program there is no response on the board. Additionally, when I switch from synplify to LSE, I see an error in the project
<img width="734" alt="image" src="https://github.com/stnolting/neorv32-setups/assets/63076423/81c71d7e-5115-49e9-8955-583aa93bab96">
I'm also still getting the read error in the programmer. Clicking the "Detect Cable" shows that it detects the FTDI on the board. I have tried different ports but no change.

**Comment by stnolting** [comment]
Hmm, seems like the PLL IP is broken. Did you regenerate the default IP module or did you recreate it?

When you **recreate** it make sure to enable the `lock_o` port, which seems to be missing on your side (see the blue box). This port is required as it is used to drive the global reset net.

![grafik](https://github.com/stnolting/neorv32-setups/assets/22944758/42314367-9c4c-4a0b-a0ed-2da7c00d8705)

I will also rebuild the IP and update the Upduino project in here.

**Comment by ajngo** [comment]
I had been re-generating all IPs up to this point. I have tried enabling the lock_o port but I still get an error
<img width="419" alt="image" src="https://github.com/stnolting/neorv32-setups/assets/63076423/76cf64ac-4066-44bc-9f0a-4a400175900c">

**Comment by stnolting** [comment]
Strange... Seems like `neorv32_xirq.vhd` is not part of the `neorv32` library. Please check the properties of that file and adjust the library.

Can you share a screen shot of the compile order / the HDL file list from radiant?

It also seems like the top entity is not defined as such. :thinking:

I am still using Radiant 2021. I will update that to the latest version and check again.

**Comment by ajngo** [comment]
Ok so I've used the updated neorv32-setups and there are no more errors that stop from synthesizing. But I'm getting the timing errors in place and route again. Here is the HDL file list
<img width="308" alt="image" src="https://github.com/stnolting/neorv32-setups/assets/63076423/967dbd2b-9e30-4791-ac42-5de66f97ddab">
<img width="282" alt="image" src="https://github.com/stnolting/neorv32-setups/assets/63076423/bb36a9de-fba4-44e6-9fc6-30f78e7aa2e4">
<img width="305" alt="image" src="https://github.com/stnolting/neorv32-setups/assets/63076423/d5388878-aed8-4fea-a887-6bf94584f38c">

**Comment by stnolting** [comment]
> Ok so I've used the updated neorv32-setups and there are no more errors that stop from synthesizing. 

That sounds great and the file hierarchy also looks good.

> But I'm getting the timing errors in place and route again.

Can you check the timing report to see where this error comes from?

You could try to reduce the PLL's output frequency to relax the timing. I will also update the Radiant project here with a reduced PLL output.

**Comment by ajngo** [comment]
After your latest commit there are no more issues with synthesizing, including timing errors. However, when I upload the core the bootloader doesn't work. I'm still using the default code so no changes.

**Comment by stnolting** [comment]
That sounds great! Are you using LSE or Synplify for synthesis?

> However, when I upload the core the bootloader doesn't work. I'm still using the default code so no changes.

This is very weird. I do not have a Upduino here, but my setup for the icebreaker board that features the same FPGA is working without problems.

I think I'll upload my setup to this repository as well. Hopefully, you can use that as starting point for your Upduino setup (basically, you just need to adapt the UART pins, right?! 🤔).

**Comment by stnolting** [comment]
I've managed to get an older UPduino v2 which is backwards compatible to the v3 version. I was **not** able to make the processor run on that board even though synthesis and everything went fine. So I did some experiments and I found out that the latest Radiant version (2023) seems to have some issues...

However, with Radiant version **2022.1** everything works fine :tada:
This older version can be downloaded from Lattice software archive: https://www.latticesemi.com/Support/SoftwareArchive

I will update the UPduino_v3 project here and I will also add my iCEbreaker setup. It would be very great if you could check the updated setup on your side (with Radiant 2022.1). :wink:

**Comment by ajngo** [comment]
Hi, thanks again for all of your help. The updated setup (2022.1) works perfectly out of the box. Something interesting to note is that I tried using the Lattice LSE as well and it gives a timing error, but when I continue, it programs with no issues.

**Comment by ajngo** [comment]
My next question is how would I go about changing the example project to configure an RV32E processor instead, if at all possible? I'm hoping that I can just edit the upduino project since that's the board I'm using, but what is your recommendation? I have already tried adding the extension to the upduino_top file and editing march and mabi in the common makefile. Is there something else that I need to change? I know that the user guide explains a different process from the example project. Any tips in the right direction would be appreciated. :)

**Comment by stnolting** [comment]
Basically, you are right. You need to
1. have a `rve`-compatible tooolchain that can emit code for a `rv32e*` architecture (like https://github.com/stnolting/riscv-gcc-prebuilt/releases/tag/rv32e-231223)
2. set the machine architecture and machine application binary interface accordingly in the makefile (`MARCH=rv32e`, `MABI=ilp32e`)
3. enable the CPU's `E` extension via the according generic (`CPU_EXTENSION_RISCV_E`)

But using `RVE` on a Lattice ice40 FPGA does not makes much sense (unfortunately). The embedded block RAMs have a fixed size of 4k bits and provide individual write and read ports. If the RAMs are configured in their "widest" mode (256x16-bit) we need two RAMs in parallel to have a word-width of 32-bit. Plus we need two 2x-RAM combos in parallel for providing two independent read ports for `rs1` and `rs2`. So we need a total of 4 RAMs.

If we implement less registers (`E` ISA extension = just 16 instead of 32 registers) we still need the same amount of RAMs as as the RAMs only support 256x16-bits in their "widest" mode.

**Comment by ajngo** [comment]
I see, that's very interesting. But I should still be able to put your RVE core onto an ice40 fpga correct? I have done the steps you listed (assuming the Radiant software can serve as a replacement for RVE toolchain) but the bootloader isn't working.

**Comment by stnolting** [comment]
> I see, that's very interesting. But I should still be able to put your RVE core onto an ice40 fpga correct?

That's right. It just won't bring any area benefits (and might reduce overall performance).

> I have done the steps you listed (assuming the Radiant software can serve as a replacement for RVE toolchain) but the bootloader isn't working.

The default bootloader image is compiled using the **default** makefile configuration (https://github.com/stnolting/neorv32/blob/main/rtl/core/neorv32_bootloader_image.vhd#L4)

```
-- MARCH: default
```

which is (https://github.com/stnolting/neorv32/blob/main/sw/common/common.mk#L53-L55):

```
# CPU architecture and ABI
MARCH ?= rv32i_zicsr_zifencei
MABI  ?= ilp32
```

If you want to run the bootloader on your RV32E setup you'll need to recompile and reinstall it (with adjusted `MARCH` and `MABI` flags):

```bash
neorv32/sw/bootloader$ make clean_all bootloader
```

**Comment by ajngo** [comment]
Ok I was able to get it working. I just want to give one final thanks for all of your help!
